### ISSUE TITLE
Rabbitmq : obsolete with ConnectionFactory ?

### ISSUE THREAD

**Issue by christopheblin** [Xabaril/AspNetCore.Diagnostics.HealthChecks#480]
I've just upgraded to latest version of this nuget package and I see that I cannot do

            builder.Add(new HealthCheckRegistration(
                name: "rabbitmq",
                factory: sp => new RabbitMQHealthCheck(sp.GetService<IConnectionFactory>()),
                failureStatus: HealthStatus.Unhealthy,
                tags: new[] { "services", "bus" },
                timeout: null
            ));

However, I do not register IConnection in the dependency injection context and the only other constructor available takes Uri + Ssl.

However, since I  use ConnectionFactory with binding in options, connectionFacotry.Uri is null and so it does not work either.

Here is a repro test case to demonstrate what happens to me

```
using FluentAssertions;
using HealthChecks.RabbitMQ;
using Microsoft.Extensions.Diagnostics.HealthChecks;
using RabbitMQ.Client;
using Xunit;

namespace Service.Tests
{
    public class Tmp
    {
        [Fact]
        public async System.Threading.Tasks.Task TmpTestAsync()
        {
            //here, I normally use IConfiguration.Bind() instaed of manually specifying params
            var connectionFactory = new ConnectionFactory()
            {
                HostName = "localhost",
                Port = 5672,
                UserName = "guest",
                Password = "guest",
                VirtualHost = "/"
            };

            //that does not work because connectionFactory.Uri is null
            var healthcheck = new RabbitMQHealthCheck(connectionFactory.Uri, connectionFactory.Ssl);

            var res = await healthcheck.CheckHealthAsync(new HealthCheckContext());
            res.Status.Should().Be(HealthStatus.Healthy);
        }
    }
}

```

How am I suppose to do ?

note that you cannot do sp.GetService<IConnectionFactory>().CreateConnection() because it thows an exception (instead of reporting Unealthy)

Thanks for any help

**Comment by unaizorrilla** [comment]
Hi @christopheblin 

We remove the IConnectionFactory because this option don't reuse connections on healtcheck and this will be a problem a long time!

The way to register a re-use connection with connection factory is:

```csharp
        services
        .AddSingleton<IConnection>(sp=>
        {
            var factory = new ConnectionFactory()
            {
                Uri = new Uri("amqps://user:pass@host/vhost"),
                AutomaticRecoveryEnabled = true
            };

            return  factory.CreateConnection();
        })
        .AddHealthChecks()
        .AddRabbitMQ();
```

**Comment by christopheblin** [comment]
Hi @unaizorrilla and thanks for the answer

With this code, if rabbitmq is not available on startup, it will throw an exception and application will never start (thus, never answering the healthchecks and, in kubernetes, restart the pod after 30 seconds). When you have 100 pods, this is not exactly what you want...

What I want is that the app starts and waits for rabbit. I have some code around this problem with a try/catch and a timer (this is why I do not have the IConnection available in DI). 

If someone tries a web request that does not need rabbit, it works. If it need rabbit, it fails.

For this case, the goal of healthchecks is to answer Unhealthy if the connection can not be established (and this was working before)

I dont see why it is a problem to open a connection and then close it for each healthcheck, can you explain please ?

**Comment by unaizorrilla** [comment]
Well, in this case, when your dependencies are not ready at startup ( not common case imho ) you can use different registration methods like

https://github.com/Xabaril/AspNetCore.Diagnostics.HealthChecks/blob/59ebfb58512dfefff5e46db20359ed96745b28a7/src/HealthChecks.RabbitMQ/DependencyInjection/RabbitMQHealthCheckBuilderExtensions.cs#L27

or 

https://github.com/Xabaril/AspNetCore.Diagnostics.HealthChecks/blob/59ebfb58512dfefff5e46db20359ed96745b28a7/src/HealthChecks.RabbitMQ/DependencyInjection/RabbitMQHealthCheckBuilderExtensions.cs#L76

**Comment by christopheblin** [comment]
This is what I described in the issue : I am not using the Uri but configuring an IConnectionFactory with Host, Port, ... (thus Uri is null).

Also, I looked at the code, using the Uri just recreate a ConnectionFactory to create a IConnection (without checking exceptions)

=> Why directly passing IConnectionFactory is not supported anymore when passing Uri is supported with exactly the same problem of opening a new connectio ?
=> Why exceptions are not catched when calling CreateConnection (which has the bad effect that /ready answers with a status code 500) ?

Otherwise, what is the best way to achieve a healthcheck that answers unehealthy when the connection can not be opened ?

Again, thanks for your time

**Comment by christopheblin** [comment]
Here is what I've done, in case it may help someone

```
    public class MyRabbitMQHealthCheck : IHealthCheck
    {
        private readonly ConnectionFactory connectionFactory;

        public MyRabbitMQHealthCheck(ConnectionFactory connectionFactory)
        {
            this.connectionFactory = connectionFactory;
        }

        public Task<HealthCheckResult> CheckHealthAsync(HealthCheckContext context, System.Threading.CancellationToken cancellationToken = default)
        {
            try
            {
                using var c = connectionFactory.CreateConnection();
                using var m = c.CreateModel();
                return Task.FromResult(HealthCheckResult.Healthy());
            }
            catch (Exception ex)
            {
                return Task.FromResult(new HealthCheckResult(context?.Registration?.FailureStatus ?? HealthStatus.Unhealthy, exception: ex));
            }
        }
    }
```

Then I do: 

```
          builder.Add(new HealthCheckRegistration(
                name: "rabbitmq",
                factory: sp => new MyRabbitMQHealthCheck(sp.GetService<IConnectionFactory>()),
                failureStatus: HealthStatus.Degraded,
                tags: new[] { "services", "bus" },
                timeout: null
            ));
```

I hope this does not break anything ...

**Comment by unaizorrilla** [comment]
IConnectionFactory is not supported ( obsolete ) because this promote not reusing the IConnection, if you read the rabbitmq client doc, IConnection is created to be a singleton. If inject the IConnectionFactory on rabbit healthcheck a new connection will be created each check iteration.

I tested your comment about startup fail if rabbit is not ready yet and in my sample is not true, if you use something like

```csharp
services.AddHealthChecks()
             .AddRabbitMQ(rabbitConnectionString: "amqp://localhost:5672", name: "rabbit1")
```

there is no problem if rabbit is not working, healthcheck is unhealthy but not unhandled exception is thrown. The only problem is not connection re-use,  info about rabbit uris https://www.rabbitmq.com/uri-spec.html. 

I tested also using the IConnection and is not throw on startup, throw on check health meanwhile the broker does not respond, if the broker response, the iconnection is created and because AutomaticRecoverEnabled is true working also well on transient failures of rabbit

```csharp
           services.AddSingleton<IConnection>(sp =>
            {
                var connectionFactory = new ConnectionFactory()
                {
                    Uri = new Uri("amqp://localhost:5672"),
                    AutomaticRecoveryEnabled = true,
                    UseBackgroundThreadsForIO = true
                };      
                return  connectionFactory.CreateConnection();
            });

            services.AddHealthChecks()
                .AddRabbitMQ(name: "rabbit1");
```

In your scenario, pods with readiness/liveness, the problem is the same because in both healthz does not response 200 ok and K8S will try to restart pods!

**Comment by unaizorrilla** [comment]
> 
> 
> Here is what I've done, in case it may help someone
> 
> ```
>     public class MyRabbitMQHealthCheck : IHealthCheck
>     {
>         private readonly ConnectionFactory connectionFactory;
> 
>         public MyRabbitMQHealthCheck(ConnectionFactory connectionFactory)
>         {
>             this.connectionFactory = connectionFactory;
>         }
> 
>         public Task<HealthCheckResult> CheckHealthAsync(HealthCheckContext context, System.Threading.CancellationToken cancellationToken = default)
>         {
>             try
>             {
>                 using var c = connectionFactory.CreateConnection();
>                 using var m = c.CreateModel();
>                 return Task.FromResult(HealthCheckResult.Healthy());
>             }
>             catch (Exception ex)
>             {
>                 return Task.FromResult(new HealthCheckResult(context?.Registration?.FailureStatus ?? HealthStatus.Unhealthy, exception: ex));
>             }
>         }
>     }
> ```
> 
> Then I do:
> 
> ```
>           builder.Add(new HealthCheckRegistration(
>                 name: "rabbitmq",
>                 factory: sp => new MyRabbitMQHealthCheck(sp.GetService<IConnectionFactory>()),
>                 failureStatus: HealthStatus.Degraded,
>                 tags: new[] { "services", "bus" },
>                 timeout: null
>             ));
> ```
> 
> I hope this does not break anything ...

As I say on my latest response, you can use obsolete method or your custom impl, but this is not recomended (is similar to your custom implementation) because leak connections and cause problems on long runing services.
Sincerely, try to use the IConnection, start the services without the dependencies ready is not common case and at the end, the issue is very similar, in both cases health can respond 200 ok and orchestrator try to restart the service!

I close the issue!

**Comment by changhuixu** [comment]
Hi @christopheblin , I agree with what you that the current library is not friendly. I came across the same issue that rabbitmq is not available when application starts, which is common in containerized applications because rabbitmq takes a few minutes to spin up.

The current library requires an instance of `IConnection` injected into the health check. However, when the rabbitmq server is not available, the DI registration simply fails and the whole application won't start at all.

Another constructor takes `Uri` and `SslOptions`, seems an overkill and cumbersome. Why don't take the `ConnectionFactory` directly? I see the code recreates a connection factory in the method `EnsureConnection()`. Is it really needed?

Another version to the one that @christopheblin wrote is as follows.
```csharp
public class RabbitMQHealthCheck : IHealthCheck
{
    private readonly ConnectionFactory _connectionFactory;
    private IConnection _connection;

    public RabbitMQHealthCheck(ConnectionFactory connectionFactory)
    {
        _connectionFactory = connectionFactory;
    }
    public Task<HealthCheckResult> CheckHealthAsync(HealthCheckContext context, CancellationToken cancellationToken = new CancellationToken())
    {
        try
        {
            if (_connection == null)
            {
                _connection = _connectionFactory.CreateConnection();
            }
            using var _ = _connection.CreateModel();
            return Task.FromResult(HealthCheckResult.Healthy());
        }
        catch (Exception e)
        {
            return Task.FromResult(HealthCheckResult.Unhealthy(e.Message));
        }
    }
}
```
NOTE: ASP.NET Core has an Unhealthy code already, so I use `HealthCheckResult.Unhealthy(e.Message)` for connection failures.

Then in Startup.cs, register them as
```csharp
  services.AddSingleton(connectionFactory);
  services.AddHealthChecks().AddCheck<RabbitMQHealthCheck>("rabbitmq");
```

In this way, the DI registration doesn't need an `IConnection`, thus won't cause the whole app failing to start.

The goal is to reduce new `IConnection` instances. However, it appears that at least 2 `IConnection` instances are required. One for health check, the other for the actual exchange.

Hope this will help someone bump into the same problem. 
@unaizorrilla Comments are welcome. Thanks.

**Comment by unaizorrilla** [comment]
Hi @changhuixu

 - Let me check to add ConnectionFactory again and the limitations!!
 - Unhealthy(e.Message) is not honorign the FailureStatus registered by the user

**Comment by unaizorrilla** [comment]
Hi @changhuixu 

Can you add review on this?

https://github.com/Xabaril/AspNetCore.Diagnostics.HealthChecks/pull/544

**Comment by unaizorrilla** [comment]
Merged #544 When a release build finished a new package 3.1.2 for rabbitmq will be available!

**Comment by changhuixu** [comment]
Thank you

**Comment by unaizorrilla** [comment]
Closed with #587

**Comment by shawndutton-unitas** [comment]
> Hi @christopheblin , I agree with what you that the current library is not friendly. I came across the same issue that rabbitmq is not available when application starts, which is common in containerized applications because rabbitmq takes a few minutes to spin up.
> 
> The current library requires an instance of `IConnection` injected into the health check. However, when the rabbitmq server is not available, the DI registration simply fails and the whole application won't start at all.
> 
> Another constructor takes `Uri` and `SslOptions`, seems an overkill and cumbersome. Why don't take the `ConnectionFactory` directly? I see the code recreates a connection factory in the method `EnsureConnection()`. Is it really needed?
> 
> Another version to the one that @christopheblin wrote is as follows.
> 
> ```cs
> public class RabbitMQHealthCheck : IHealthCheck
> {
>     private readonly ConnectionFactory _connectionFactory;
>     private IConnection _connection;
> 
>     public RabbitMQHealthCheck(ConnectionFactory connectionFactory)
>     {
>         _connectionFactory = connectionFactory;
>     }
>     public Task<HealthCheckResult> CheckHealthAsync(HealthCheckContext context, CancellationToken cancellationToken = new CancellationToken())
>     {
>         try
>         {
>             if (_connection == null)
>             {
>                 _connection = _connectionFactory.CreateConnection();
>             }
>             using var _ = _connection.CreateModel();
>             return Task.FromResult(HealthCheckResult.Healthy());
>         }
>         catch (Exception e)
>         {
>             return Task.FromResult(HealthCheckResult.Unhealthy(e.Message));
>         }
>     }
> }
> ```
> 
> NOTE: ASP.NET Core has an Unhealthy code already, so I use `HealthCheckResult.Unhealthy(e.Message)` for connection failures.
> 
> Then in Startup.cs, register them as
> 
> ```cs
>   services.AddSingleton(connectionFactory);
>   services.AddHealthChecks().AddCheck<RabbitMQHealthCheck>("rabbitmq");
> ```
> 
> In this way, the DI registration doesn't need an `IConnection`, thus won't cause the whole app failing to start.
> 
> The goal is to reduce new `IConnection` instances. However, it appears that at least 2 `IConnection` instances are required. One for health check, the other for the actual exchange.
> 
> Hope this will help someone bump into the same problem. @unaizorrilla Comments are welcome. Thanks.

Thank you for posting this as it was a great help to me.

I ran into this when upgrading the package as the supplying the connection factory is no longer supported.  The solutions in the readme all grenade with an exception if the endpoint isn't available instead of returning unhealthy as I wanted.

This above seems to work perfectly.

Thanks again!
